### ISSUE TITLE
Sonic Robo Blast 2 segment violation

### ISSUE THREAD

**Issue by Zeritar** [clementgallet/libTAS#358]
Been having this issue for a while, game occasionally (9/10 times) crashes at a specific point. Noticed the option to disable audio in 1.4 and upon doing so, it doesn't crash anymore. There's some stuff spammed in the console right before the crash.

```
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61342
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61342
[libTAS f:2112] Thread 28379 (main) SDL_Delay call - sleep for 1 ms.
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_WasInit with flags 512
[libTAS f:2112] Thread 28379 (main) SDL_JoystickUpdate call.
[libTAS f:2112] Thread 28379 (main) SDL_PollEvent call.
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_GetTicks call - returning 61343
[libTAS f:2112] Thread 28379 (main) SDL_RenderPresent call.
[libTAS f:2112] Thread 28395        SDL_LockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_UnlockAudioDevice call.
[libTAS f:2112] Thread 28395        Remove SDL Timer.
[libTAS f:2112] Thread 28395        SDL_LockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_UnlockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_LockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_UnlockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_LockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_UnlockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_LockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_UnlockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_LockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_UnlockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_LockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_UnlockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_LockAudioDevice call.
[libTAS f:2112] Thread 28395        SDL_UnlockAudioDevice call.

Process killed by signal: SIGSEGV - segment violation (core dumped)
```

**Comment by clementgallet** [comment]
Does this happen when you are doing a specific action in the game? I've played a bit and didn't encounter any crash. Alternatively, do you have a movie file that triggers the crash?

**Comment by Zeritar** [comment]
Right, sorry, I should have been more specific. It happens close to the end of Castle Eggman Act 1.

We did some digging last night and actually managed to find the cause of the crash. When you reach this checkpoint (2nd to last),
![crash](https://user-images.githubusercontent.com/43818834/87776764-8ee76d80-c828-11ea-91ec-66b8d6725d5d.png)
there is a trigger that causes the music to slowly fade out in preparation for the next part. Once the music is completely faded out, the game crashes. We tested this by lingering around the checkpoint instead of proceeding and it still crashed. I then went and edited the map to disable the music fade and that stopped the crash from happening

Unfurtunately the only movie files I have were made with 1.3.5 and they desync with 1.4.0 so it won't get you to the crash point, but I have an in-game demo that crashes when run with libTAS:
[amytest.zip](https://github.com/clementgallet/libTAS/files/4937218/amytest.zip)

amytest.lmp goes in ~/.srb2 and is played with the command-line argument `-playdemo amytest`

**Comment by clementgallet** [comment]
Thanks! Backtrace:

```
Thread 1 "lsdl2srb2" received signal SIGSEGV, Segmentation fault.
0x00007ffff65b8828 in mdct_backward () from /usr/lib/x86_64-linux-gnu/libvorbis.so.0
(gdb) bt
#0  0x00007ffff65b8828 in mdct_backward () at /usr/lib/x86_64-linux-gnu/libvorbis.so.0
#1  0x00007ffff65c953f in  () at /usr/lib/x86_64-linux-gnu/libvorbis.so.0
#2  0x00007ffff6697654 in  () at /usr/lib/x86_64-linux-gnu/libvorbisfile.so.3
#3  0x00007ffff6699fcd in ov_read_filter () at /usr/lib/x86_64-linux-gnu/libvorbisfile.so.3
#4  0x00007ffff669a3e1 in ov_read () at /usr/lib/x86_64-linux-gnu/libvorbisfile.so.3
#5  0x00007ffff7d35930 in  () at /usr/lib/x86_64-linux-gnu/libSDL2_mixer-2.0.so.0
#6  0x00007ffff7d31aa4 in music_pcm_getaudio () at /usr/lib/x86_64-linux-gnu/libSDL2_mixer-2.0.so.0
#7  0x00007ffff7d31c5e in music_mixer () at /usr/lib/x86_64-linux-gnu/libSDL2_mixer-2.0.so.0
#8  0x00007ffff7d2f97e in  () at /usr/lib/x86_64-linux-gnu/libSDL2_mixer-2.0.so.0
#9  0x00007ffff7f15487 in libtas::AudioSource::mixWith(timespec, unsigned char*, int, int, int, int, float)
    (this=0x555563b2f8b0, ticks=..., outSamples=0x55555c863b00 "", outBytes=<optimized out>, outBitDepth=16, outNbChannels=2, outFrequency=44100, outVolume=<optimized out>)
    at audio/AudioSource.cpp:334
#10 0x00007ffff7f12bb9 in libtas::AudioContext::mixAllSources(timespec) (this=0x7ffff7fc4c80 <libtas::audiocontext>, ticks=...) at /usr/include/c++/9/bits/shared_ptr_base.h:1020
#11 0x00007ffff7ef7f60 in libtas::frameBoundary(bool, std::function<void ()>, libtas::RenderHUD&) (drawFB=drawFB@entry=true, draw=..., hud=...) at frame.cpp:210
#12 0x00007ffff7f6a819 in libtas::SDL_RenderPresent(SDL_Renderer*) (renderer=<optimized out>) at /usr/include/c++/9/new:174
#13 0x000055555572ff90 in I_FinishUpdate () at sdl/i_video.c:1241
#14 0x00005555555a0fe8 in D_Display () at d_main.c:679
#15 0x00005555555a1f15 in D_Display () at d_main.c:816
#16 D_SRB2Loop () at d_main.c:816
#17 0x00005555555a05f7 in main (argc=<optimized out>, argv=<optimized out>) at sdl/i_main.c:258
```

**Comment by clementgallet** [comment]
Fixed in b3140f4ff4850489a4906934fe9d299536b6a91a
